### ISSUE TITLE
Creating a kubernetes client should adapt if the k8s server is openshift

### ISSUE THREAD

**Issue by davsclaus** [fabric8io/kubernetes-client#383]
Talked to jimmi on chat today.

We have a Camel kubernetes component that uses this client. It creates a client as

```
        Config conf = builder.build();
        return new DefaultKubernetesClient(conf);
```

If you use the client and specify a valid username/password/master url, then the client cannot login.
You get errors like in this gist: https://gist.github.com/davsclaus/9325dbf80b359b746825f62aa4a4281b

However if you create the client as
```
        Config conf = builder.build();
        return new DefaultOpenShiftClient(conf);
```

Then the client works with just username/password for authentication.

Jimmi mention that creating the kubernetes client should be able to adapt to the OS environment automatic.

**Comment by davsclaus** [comment]
This is part of a camel-kubernetes component where the client is created at
https://github.com/apache/camel/blob/kube-lb/components/camel-kubernetes/src/main/java/org/apache/camel/component/kubernetes/processor/KubernetesServiceProcessor.java#L138

To allow calling services from Camel where you want to discover and do client side load balancing.
https://issues.apache.org/jira/browse/CAMEL-9683

**Comment by davsclaus** [comment]
An unit test in camel-kubernetes where I setup the details to use for the client as
https://github.com/apache/camel/blob/kube-lb/components/camel-kubernetes/src/test/java/org/apache/camel/component/kubernetes/processor/ToServiceRouteTest.java

**Comment by iocanel** [comment]
@davsclaus: So what's different between the vanilla kubernetes client?

Openshift provides a way of creating an oauth token given a username and password. So even if u provide username and password in the end you authenticate using an oauth token (so it pretty much does what ``oc login`` would do.

The kubernetes client on the other hand doesn't have something like this (at least I am not aware of it). So when specifying username and password it tries plain old basic authentication. 

The question here is. ``is your target environment supposed to work with basic auth?``

**Comment by iocanel** [comment]
I see two issues here:

- Basic authentication is not handled correctly by our client (I am on it).
- Assuming you're using the vagrant openshift installer, your env is not expected to work with basic auth.

"automatic" adaptation would help here, but we don't have such thing, nor its something we can easily implement. Mostly because OpenshiftClient is not instance of KubernetesClient, so we can't just detect environment and serve a different implementation.

**Comment by jimmidyson** [comment]
But we could add the oauth interceptor in if we identify that we're running against OpenShift?

**Comment by iocanel** [comment]
To detect if environment is openshift we need to get the root paths, which is something we can do even if we are not authenticated, so technically yeah we can do that.

**Comment by iocanel** [comment]
If OpenshiftClient extended KubernetesClient we could introduce the auto adaptation concept. That could work somehow like that:

    KunernetesClientFactory factory = new KubernetesClientFactory();
    KubernetesClient client = factory.create(config);

The Factory could look up for all available adapters and if an adapter had a property say ``autoAdapt`` true it could try and automatically adapt.

An other approach if we wanted to avoid the factory, could be something like:

    KubernetesClient client = new EnvironmentAwareKubernetesClient(config);

The more I think of it the more I like the second approach.

Anyway the only obstacle here is that OpenshiftClient doesn't extend KubernetesClient.
Why? 
At some point @jstrachan raised issue #123 and the idea was to be able change the default namespace. But since we didn't want to maintain state on the client we created a new client instance every time we change the namespace, as described in the issue. 

The catch here is that the inNamespace() method should return an instance of KubernetesClient when used by a KubernetesClient, but should return an OpenshiftClient when used by an OpenshfitClient. That pretty much rules out inherience as you can't have a method with different return values in the same hierarchy.

So the options here were to expose generics to the user, which was a no go, or to push common stuff between the two interfaces one level up (in an interface called KubernetesDSL) and then have 2 clients that don't inherit each other.

We went with the later option. And here we are :-)

What I propose to do?

Move the inNamespace() methods from the KubernetesClient to NamespaceableKubernetesClient and NamespaceableOpenshiftClient and let everyone that wants to use that feature explicitly specify it.

Phew! Thoughts?

**Comment by jimmidyson** [comment]
Can't we just leave it the way it is & use the `OpenShiftExtensionAdapter` logic (if available on classpath) to identify whether it's an OpenShift server or not, adding oauth interceptor as required? This would keep it exactly as it is with a pretty simple change?

**Comment by iocanel** [comment]
@jimmidyson, @davsclaus: Here's how it could look like: 

https://github.com/fabric8io/kubernetes-client/pull/385

**Comment by jimmidyson** [comment]
Fixed by #385
